jormungandr/src/blockchain/chain.rs

```diff
@@ -66,57 +66,97 @@ use tokio::stream::StreamExt;
-            description("Block0 is not yet in the storage")
-        }
+    #[error("")]
+    Block0 {
+        #[from]
+        #[source]
+        source: Block0Error,
+    },
 
-        MissingParentBlock(hash: HeaderHash) {
-            description("missing a parent block from the storage"),
-            display(
-                "Missing a block from the storage. The node was recovering \
-                 the blockchain and the parent block '{}' was not \
-                 already stored",
-                hash,
-            ),
-        }
+    #[error("Error while creating the initial ledger out of the block0")]
+    Block0InitialLedgerError,
 
-        NoTag (tag: String) {
-            description("Missing tag from the storage"),
-            display("Tag '{}' not found in the storage", tag),
-        }
+    #[error("Block0 already exists in the storage")]
+    Block0AlreadyInStorage,
 
-        BlockHeaderVerificationFailed (reason: String) {
-            description("Block header verification failed"),
-            display("The block header verification failed: {}", reason),
-        }
+    #[error("Block0 is not yet in the storage")]
+    Block0NotAlreadyInStorage,
+
+    #[error(
+        "Missing a block from the storage. The node was recovering \
+          the blockchain and the parent block '{hash}' was not \
+          already stored"
+    )]
+    MissingParentBlock { hash: HeaderHash },
+
+    #[error("Tag '{tag}' not found in the storage")]
+    NoTag { tag: String },
+
+    #[error("The block header verification failed: {reason}")]
+    BlockHeaderVerificationFailed { reason: String },
 
-        BlockNotRequested (hash: HeaderHash) {
-            description("Received an unknown block"),
-            display("Received block {} is not known from previously received headers", hash)
+    #[error("Received block {hash} is not known from previously received headers")]
+    BlockNotRequested { hash: HeaderHash },
+
+    #[error("Block cannot be applied on top of the previous block's ledger state")]
+    CannotApplyBlock,
+
+    #[error("{message}")]
+    Custom { message: String },
+
+    #[error("{error}")]
+    Chained {
+        error: Box<Self>,
+        #[source]
+        chain: Box<dyn std::error::Error + Send + 'static>,
+    },
+}
+
+impl Error {
+    pub fn with_chain<E, M>(e: E, message: M) -> Self
+    where
+        M: Into<Self>,
+        E: std::error::Error + Send + 'static,
+    {
+        Self::Chained {
+            error: Box::new(message.into()),
+            chain: Box::new(e),
         }
+    }
+}
 
-        CannotApplyBlock {
-            description("Block cannot be applied on top of the previous block's ledger state"),
+impl<'a> From<&'a str> for Error {
+    fn from(message: &'a str) -> Self {
+        Self::Custom {
+            message: message.to_owned(),
         }
     }
 }
 
+impl From<String> for Error {
+    fn from(message: String) -> Self {
+        Self::Custom { message }
+    }
+}
```

Comment:
I think these two could be unified into `impl<T: Into<String>> From<T> for Error`, but this is finessing it.

Maybe even better to replace these with an `Error::custom` constructor to be able to audit uses, this looks like a stopgap when someone could not be bothered to create another error variant.